trade-finance-manager-api/src/v1/controllers/user/user.controller.upsert-tfm-user-from-entra-id-user.test.js

```diff
@@ -0,0 +1,54 @@
+import { when, resetAllWhenMocks } from 'jest-when';
+import { mapUserData } from './helpers/mapUserData.helper';
+import { UserService } from '../../services/user.service';
+import { upsertTfmUserFromEntraIdUser } from './user.controller';
+
+jest.mock('../../services/user.service', () => ({
+  __esModule: true,
+  UserService: {
+    upsertUserFromEntraIdUser: jest.fn(),
+  },
+}));
+
+jest.mock('./helpers/mapUserData.helper', () => ({
```

Comment:
Do we need to mock this?

Reply:
the tldr of this is that if it isn't mocked, we'd have to maintain an accurate `UserService.upsertUserFromEntraIdUser` response -- which is not ideal at all (as since it's not from the repo layer, it could have mappings). I'd like to set the idea that in tests specifically like this (orchestration functions that call async, and do something with an async response), mocking is used as it's less maintainence overall -- but in tests like the others I've done we can do that neat way demonstrated in this PR and by yourself.

Reply:
I think it'd be much simpler to do this?

```ts
it('should return the mapped user', async () => {
  const result = await makeRequest();

  const expected = mapUserData(upsertedUserResponse);

  expect(result).toEqual(expected);
});
```

Since the actual implemention simply calls `mapUserData` with the upsert response

And then we don't need the following:

```ts
jest.mock('./helpers/mapUserData.helper', () => ({
  mapUserData: jest.fn(),
}));

when(jest.mocked(mapUserData)).calledWith(upsertedUserResponse).mockReturnValue(mappedUserDetails);
```

Reply:
I'm a little confused here.

We will have to mock `upsertedUserResponse` as an accurate `UserService.upsertedUserResponse` response for this test to be accurate, and have `mapUserData` behave in a way that is meaningful for this test.

This has a problem if `UserService` response changes -- this test wont fail as a result. We will have, in this test, an out of date `upsertedUserResponse` object that is still passing tests, and could be relied on by people as an accurate `upsertedUserResponse` when it is out of date.

We'd have to conciously and proactively look to maintain a valid `UserService.upsertedUserResponse` whenever we update `UserService.upsertedUserResponse`. This would mean every time we update this function, we would instinctively have to look through where it is called by all other functions, and update these.

If that's a pattern we wish to propergate, we should look to create a test double objects library, as if any async service responses change, we will need to ensure we remember to update the test double object.

---

So for this example -- 

```
describe('user controller', () => {
  describe('upsertTfmUserFromEntraIdUser', () => {
    const entraIdUser = 'anEntraIdUser';
    const auditDetails = 'auditDetails';

    let upsertedUserResponse
    let mappedUserDetails

    beforeEach(() => {
      jest.resetAllMocks();
      resetAllWhenMocks();

      upsertedUserReponse = aValidUpsertUserReponse();
      mappedUserDetails = mapUserData(upsertedUserResponse);
      when(jest.mocked(UserService.upsertUserFromEntraIdUser)).calledWith({ entraIdUser, auditDetails }).mockResolvedValue(upsertedUserResponse);
      when(jest.mocked(mapUserData)).calledWith(upsertedUserResponse).mockReturnValue(mappedUserDetails);
    });

    it('should upsert user in the database', async () => {
      await makeRequest();

      expect(UserService.upsertUserFromEntraIdUser).toHaveBeenCalledWith({ entraIdUser, auditDetails });
    });
    
    ....

    async function makeRequest() {
      return await upsertTfmUserFromEntraIdUser({ entraIdUser, auditDetails });
    }

    // This function is tying the unit test of an orchestration controller to the implimnentation of the user service it is calling
    // If the user service return value updates, this will not be automatically updated
    // and is prone to go out of sync with the actual values
    // leading to false confidence that the tests pass with the correct values
    async function aValidUpsertUserResponse() {
        return {
            _id: new ObjectId()
            azureOid: 'an-azure-oid',
            email: 'an-email',
            username: 'a-username',
            teams: ['BUSINESS_SUPPORT'],
            timezone: 'Europe/London',
            firstName: 'a-first-name',
            lastName: 'a-last-name',
            lastLogin: Date.now(),
        }
    }
  });
});
```

Reply:
I think a prime example of it falling out of sync is all our tfm users in our unit tests -- a search for `user` shows how these are in varying degrees of out-of-date-ness

Reply:
I'm not sure I really agree with this as I'd argue the smallest possible unit to test in a controller is it's orchestration of functions -- but as not to be dogmatic, I'll add a test double for this

Reply:
Added a test double -- as a result of this and to ensure that this ticket isn't adding tech debt in this area, a lot of changes have been made to the current ticket. Sorry in advance!

A summary explanation of why there's so much changed:
- Extracted the creation of the `UserService` `upsertUserFromEntraIdResponse` into a new `userServiceMockResponses` file
- Created a new exported type for `upsertUserFromEntraIdResponse` in `UserService` so that if this type is updated, we will get typeErrors in our tests
- Utilised the existing `tfmUser` creator from DTFS-Central-API for `upsertUserFromEntraIdResponse` as apposed to copying this file to avoid new tech debt
- This required moving this into the common libs -- So I've updated all DTFS-Central-API references in that regard
- As this move continues the pattern inside our libs common for creation of test mock data inside `test-helpers/mock-data`, I've moved other creation of test mock data there from this PR instead of schemas --such as `create-user-request` etc
- As this now is completely separate from the schema, we should call these objects `aCreateTfmUserRequest` as apposed to `aCreateUserRequest` as this would be disingenuous, and allow for accidental misuse in Portal etc, as it would be `import {aCreateUserRequest} from '@ukex/dtfs2-common`, returning a `CreateUserRequest` type, which is highly confusing and also not keeping in pattern with `aTfmUser`
- As a result, I've updated all types/schemas in this PR to reference `TfmUser` as apposed to just `User`
- To avoid creating tech debt I've continued this name change wherever we are referencing these types (as we require params to be called i.e. `UpsertTfmUserFromEntraIdUserParams` now instead of `UpsertUserFromEntraIdUserParams`, the function is called `upsertTfmUserFromEntraUserId` instead of `upsertUserFromEntraUserId` to avoid confusion.